Backport of:

From 69490f5cffbda612e15a2985699455bb0b45e276 Mon Sep 17 00:00:00 2001
From: cristy <cristy@aa41f4f7-0bf4-0310-aa73-e5a19afd5a74>
Date: Tue, 2 Dec 2014 02:11:59 +0000
Subject: [PATCH] Fix handling of corrupted of psd, sun and xpm file

Fix a heap overflow in psd file.

Bail early in xpm file.

Fix error handling in sun file.

git-svn-id: https://subversion.imagemagick.org/subversion/ImageMagick/branches/ImageMagick-6@17149 aa41f4f7-0bf4-0310-aa73-e5a19afd5a74
origin: http://trac.imagemagick.org/changeset/17149
---
 coders/psd.c        |  4 ++--
 coders/sun.c        |  4 ++--
 coders/xpm.c        | 33 +++++++++++++++------------------
 config/english.xml  |  6 +++---
 config/francais.xml |  6 +++---
 5 files changed, 25 insertions(+), 28 deletions(-)

Index: imagemagick-6.6.9.7/coders/sun.c
===================================================================
--- imagemagick-6.6.9.7.orig/coders/sun.c	2016-11-15 12:41:05.435696664 -0500
+++ imagemagick-6.6.9.7/coders/sun.c	2016-11-15 12:41:05.431696615 -0500
@@ -305,9 +305,9 @@
     sun_info.maplength=ReadBlobMSBLong(image);
     if ((sun_info.type != RT_STANDARD) && (sun_info.type != RT_ENCODED) &&
         (sun_info.type != RT_FORMAT_RGB))
-      ThrowReaderException(CoderError,"ImproperImageHeader");
+      ThrowReaderException(CorruptImageError,"ImproperImageHeader");
     if ((sun_info.maptype == RMT_NONE) && (sun_info.maplength != 0))
-      ThrowReaderException(CoderError,"ImproperImageHeader");
+      ThrowReaderException(CorruptImageError,"ImproperImageHeader");
     if ((sun_info.depth == 0) || (sun_info.depth > 32))
       ThrowReaderException(CorruptImageError,"ImproperImageHeader");
     if ((sun_info.maptype != RMT_NONE) && (sun_info.maptype != RMT_EQUAL_RGB) &&
Index: imagemagick-6.6.9.7/config/english.xml
===================================================================
--- imagemagick-6.6.9.7.orig/config/english.xml	2016-11-15 12:41:05.435696664 -0500
+++ imagemagick-6.6.9.7/config/english.xml	2016-11-15 12:41:05.431696615 -0500
@@ -116,6 +116,9 @@
         <message name="ImageIsNotTiled">
           image is not tiled
         </message>
+        <message name="ImproperImageHeader">
+          improper image header
+        </message>
         <message name="IrregularChannelGeometryNotSupported">
           irregular channel geometry not supported
         </message>
@@ -249,9 +252,6 @@
           <message name="ImageTypeNotSupported">
             image type not supported
           </message>
-          <message name="ImproperImageHeader">
-            improper image header
-          </message>
           <message name="InsufficientImageDataInFile">
             insufficient image data in file
           </message>
Index: imagemagick-6.6.9.7/config/francais.xml
===================================================================
--- imagemagick-6.6.9.7.orig/config/francais.xml	2016-11-15 12:41:05.435696664 -0500
+++ imagemagick-6.6.9.7/config/francais.xml	2016-11-15 12:41:05.431696615 -0500
@@ -110,6 +110,9 @@
         <message name="ImageDoesNotHaveAAlphaChannel">
           l'image n'a pas de canal de transparence
         </message>
+        <message name="ImproperImageHeader">
+          En-tête d'image incorrect
+        </message>
         <message name="ImageIsNotTiled">
           l'image n'a pas tiled
         </message>
@@ -249,9 +252,6 @@
           <message name="ImageTypeNotSupported">
             type d'image non supporté
           </message>
-          <message name="ImproperImageHeader">
-            En-tête d'image incorrect
-          </message>
           <message name="InsufficientImageDataInFile">
             Pas assez de données d'image dans le fichier
           </message>
